### ISSUE TITLE
Modules are registered more than once when scanning assemblies

### ISSUE THREAD

**Issue by caioproiete** [autofac/Autofac#747]
When calling `RegisterAssemblyModules` to register all modules in one or more assemblies, any module that have already previously been registered, is registered again.

Example:

```
builder.RegisterModule(new MyCustomModule());

var assembliesInAppDomain = AppDomain.CurrentDomain.GetAssemblies().ToArray();
builder.RegisterAssemblyModules(assembliesInAppDomain);
```

As you can see in the snippet above, I'm already explicitly registering `MyCustomModule`, however it will get registered again by the `RegisterAssemblyModules` call, given that `MyCustomModule` is definitely somewhere in one of the assemblies loaded in the AppDomain.

I believe we should change `RegisterAssemblyModules` to skip modules already registered.

Thoughts?

**Comment by tillig** [comment]
This is also sort of "functions as designed." Since modules can be registered with parameters, there's no way to tell if you're doing this...

```csharp
// Oops - duplicate!
builder.RegisterModule<MyModule>();
builder.RegisterModule<MyModule>();
```

Or intentionally doing this...

```csharp
// Same module acts differently based on parameters
builder.RegisterModule(new MyModule("param1"));
builder.RegisterModule(new MyModule("param2"));
```

Given we can't key off of type alone, there's not much we can key off of to determine if something is truly a duplicate.

I could see potentially adding LINQ support to assembly module scanning/registration so it has support like `RegisterAssemblyTypes`, something like this:

```csharp
builder.RegisterAssemblyModules(assemblies).Where(t => t != typeof(MyModule));
```

However, we don't really have that right now. Happy to take pull requests, though.

I'll leave this issue open as an enhancement to remind us that this may be an interesting feature we could add.

In the meantime, there's not really a workaround or fix for you that I can see, other than to maybe use things like `PreserveDefaults` to avoid overriding things that have already been set, or to be more selective about the assembly scanning that happens so you don't double-register.

**Comment by tillig** [comment]
As mentioned in #748, one option might be to have an attribute-based approach for skipping module registration for a specific module, but I think the LINQ approach would offer more flexibility (and could technically allow for attribute-based handling later on).

Another workaround here might be that instead of using `RegisterAssemblyModules`, you could also just do a foreach over the set of appropriate assembly types and register them yourself. Off the top of my head (no compiler) it'd look like...

```
var modules = assemblies
    .SelectMany(a => a.GetTypes())
    .Where(t => typeof(IModule).IsAssignableFrom(t));
foreach(var module in modules)
{
  builder.RegisterModule(module);
}
```

Using that LINQ workaround it would be easy enough to exclude or include types of your own choosing. I'm guessing that what we'd offer would be roughly equivalent to that.

**Comment by caioproiete** [comment]
On detecting if a module has already been registered, what if we required modules to implement `IEquatable<IModule>`, so we can keep a registry of the modules as they are registered and can skip registration if they already exist?

I took a first stab at it on PR #752 for us to see how it could look like.

Known issues:
- This is a breaking change, no doubt. The `Module` class could provide a default implementation, but I think that could cause more issues. Leaving the method abstract at least will force users to implement the right equality checker for them;

- I have not written any test yet. Happy update the PR with tests if this approach makes sense;

- [The AppVeyor build is failing](https://ci.appveyor.com/project/alexmg/autofac-raid8/build/4.0.0.205) with error `CS0246: The type or namespace name 'ConditionalWeakTable<,>' could not be found (are you missing a using directive or an assembly reference?)`. Not sure if `ConditionalWeakTable` didn't make it to .NET Core RC2, or if it is something I forgot to update in project.json. Happy to investigate further if this approach makes sense.

**Comment by alexmg** [comment]
I don't really like the idea of forcing equality implementations or breaking existing behaviour for this rare case. What about a simple overload with a `Func<Type, bool> filter` predicate? We could even redirect the existing overload that takes a specific `Type` to this new method as a simple filter. This approach would mean some manual work is required to prevent the second module registration, but implementing equality on a module is also explicit and manual work anyway.

**Comment by tillig** [comment]
I'm also not into the requirement for equality. That's why I was thinking that LINQ support for `RegisterAssemblyModules` the way we have it for `RegisterAssemblyTypes` would be good - you could use `Distinct()` or any other lambda/predicate (e.g., looking at attributes) to filter as needed. And it would solve #748 - two birds with one stone.

Even with LINQ, we'd still not catch the situation where someone manually registers a module and then re-registers it during assembly scanning. That's not even necessarily an incorrect behavior, depending on the situation. So the value here, to be clear, is only in handling a single `RegisterAssemblyModules` call, and not in affecting global `ContainerBuilder` or module registration behavior.

**Comment by caioproiete** [comment]
Hi @alexmg and @tillig, and thanks for your comments.

Without a way of assigning some sort of identity to each module, such as with the equality impl, I don't see how we can prevent the user from registering duplicated modules. Either using assembly scanning, or explicitly registering individual modules.

Using a predicate to filter what gets registered is def. a useful feature to have, no doubt, but that doesn't help in the cases where we don't control (or don't know) modules in other assemblies.

If I publish a library on NuGet that includes Autofac modules, and the user uses assembly scanning to register modules, he/she may inadvertently register the modules from my libraries twice, without knowing. Which is what happened with the Autofac-Serilog integration (https://github.com/nblumhardt/autofac-serilog-integration/pull/10/files)

There's also cases where the user just makes a mistake and registers something twice. I'd expect to get an error on these cases, or a warning at the very least.

@tillig:
> Even with LINQ, we'd still not catch the situation where someone manually registers a module and then re-registers it during assembly scanning

From the (manual) tests I did, implementing the equality check will block any attempt to register the same module twice... Using assembly scanning, or not.

**Comment by tillig** [comment]
We [added a generic properties dictionary](https://github.com/autofac/Autofac/commit/285e407dbec5ae7480ea16ab3a753dc0e7ecb1e1) to the `ContainerBuilder` which propagates to the built `IComponentRegistry`. This is available during build/registration as well as resolution time. In nested lifetime scopes, you can modify the property dictionary just for that child scope and not affect the dictionary from the parent scope.

I think you could solve the duplicate module registration / equality comparison using this dictionary. Since we don't want to break the Module interface and we don't want to restrict modules from being registered multiple times should the app author choose, it'd be an opt-in thing.

In the module, look for a known value in the properties dictionary. If it exists, don't do any component registration. If it doesn't, add the value and do your registrations.

It would mean requiring a newer version of Autofac - one with the properties dictionary in place - but that's not much different than if we had added equality to the Module object.

Will that work for you?

**Comment by tillig** [comment]
The generic properties dictionary version of Autofac was published to NuGet as v4.2.0. [We have a unit test illustrating how to use the property bag to opt-in and avoid duplicate module registration.](https://github.com/autofac/Autofac/blob/v4.2.0/test/Autofac.Test/ModuleTests.cs#L109) Hopefully that's enough to help you address the challenge you're facing.

**Comment by cordasfilip** [comment]
@tillig I know this is an old one but I do feel like there is an issue with registering modules I get that this 
```
builder.RegisterModule<MyModule>();
builder.RegisterModule<MyModule>();
```
will register the module twice but shouldn't this 
```
var myModule = new MyModule("param1");
builder.RegisterModule(myModule);
builder.RegisterModule(myModule);
```
Do it only once? I am registering the exact same instance of the module so this should be done only once. This can be solved by adding a flag on Module class and only running load once.

Also might be nice to provide an implementation of a  singular module in the framework. I usually use 
public abstract class SingularModule : Module
    {

        public void Configure(ContainerBuilder builder) { }

        /// <summary>
        /// Runs only the first time module is registered.
        /// </summary>
        /// <param name="builder">The builder through which components can be registered.</param>
        protected virtual void InitialLoad(ContainerBuilder builder)
        {

        }

        /// <summary>
        /// Override to add registrations to the container.
        /// </summary>
        /// <param name="builder">The builder through which components can be registered.</param>
        protected virtual void EveryLoad(ContainerBuilder builder)
        {

        }

        protected sealed override void Load(ContainerBuilder builder)
        {
            if (IsInitialLoad(builder))
            {
                SetInitial(builder);
                InitialLoad(builder);
            }
            EveryLoad(builder);
        }

        private bool IsInitialLoad(ContainerBuilder componentRegistry)
        {
            return !componentRegistry.Properties.ContainsKey(GetType().AssemblyQualifiedName);
        }

        private void SetInitial(ContainerBuilder componentRegistry)
        {
             componentRegistry.Properties.Add(GetType().AssemblyQualifiedName, true);
        }
    }

Or add a way to declare that my module depends on some other modules. I use to use dependency injection for this that's how I noticed the problem for registering the same instance of a module.
